elements-fun/src/transaction.rs

```diff
@@ -23,34 +23,196 @@ use bitcoin::blockdata::script::{Instruction, Script};
 use bitcoin::hashes::Hash;
 use bitcoin::{self, Txid, VarInt};
 
-use crate::confidential;
-use crate::encode::{self, Decodable, Encodable};
+use crate::confidential::AssetCommitment;
+use crate::confidential::NonceCommitment;
+use crate::confidential::ValueCommitment;
+use crate::encode::{self, Decodable, Encodable, Error};
 use crate::issuance::AssetId;
 
+// TODO: Is `Default` really useful here?
 /// Description of an asset issuance in a transaction input
-#[derive(Copy, Clone, Debug, Default, Eq, Hash, PartialEq, PartialOrd, Ord)]
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(crate = "serde_crate")
+)]
+pub enum AssetIssuance {
+    Null(NullAssetIssuance),
+    Explicit(ExplicitAssetIssuance),
+    Confidential(ConfidentialAssetIssuance),
+}
+
+impl AssetIssuance {
+    pub fn encoded_length(&self) -> usize {
+        match self {
+            AssetIssuance::Null(inner) => inner.encoded_length(),
+            AssetIssuance::Explicit(inner) => inner.encoded_length(),
+            AssetIssuance::Confidential(inner) => inner.encoded_length(),
+        }
+    }
+}
+
+impl Default for AssetIssuance {
+    fn default() -> Self {
+        Self::Null(NullAssetIssuance::default())
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 #[cfg_attr(
     feature = "serde",
     derive(serde::Serialize, serde::Deserialize),
     serde(crate = "serde_crate")
 )]
-pub struct AssetIssuance {
+pub struct ExplicitAssetIssuance {
     /// Zero for a new asset issuance; otherwise a blinding factor for the input
     pub asset_blinding_nonce: [u8; 32],
     /// Freeform entropy field
     pub asset_entropy: [u8; 32],
     /// Amount of asset to issue
-    pub amount: confidential::Value,
+    pub amount: ExplicitValue,
     /// Amount of inflation keys to issue
-    pub inflation_keys: confidential::Value,
+    pub inflation_keys: ExplicitValue,
+}
+
+impl ExplicitAssetIssuance {
+    pub fn encoded_length(&self) -> usize {
+        32 + 32 + self.amount.encoded_length() + self.inflation_keys.encoded_length()
```

Comment:
:question: Should we use a `SECRET_KEY_LENGTH` sort of constant (that may already exist) instead of magic numbers?